src/main/java/com/connexta/jsonrpc/impl/JsonRpc.java

```diff
@@ -0,0 +1,144 @@
+    for (MethodSet methods : methodSets) {
+      builder.putAll(methods.getMethods());
+    }
+    this.defaultMethods = builder;
+    recompileMethodList();
+  }
+
+  @Override
+  public Object apply(Map<String, Object> request) {
+    return process(request);
+  }
+
+  public synchronized void bindMethods(MethodSet methodSet) {
+    dynamicMethods.add(methodSet);
+    recompileMethodList();
+  }
+
+  public synchronized void unbindMethods(MethodSet methodSet) {
+    dynamicMethods.remove(methodSet);
+    recompileMethodList();
+  }
+
+  private void recompileMethodList() {
+    Map<String, RpcMethod> newMethodList = new HashMap<>(defaultMethods);
+
+    newMethodList.putAll(
+        dynamicMethods
+            .stream()
+            .flatMap(ms -> ms.getMethods().entrySet().stream())
+            .collect(Collectors.toMap(Entry::getKey, Entry::getValue, (oldVal, newVal) -> newVal)));
+
+    methods = newMethodList;
+  }
+
+  private Object process(Map<String, Object> request) {
+    Object id = request.get(ID);
+
+    Instant start = Instant.now();
+    Object response = dispatch(request);
+    Duration duration = Duration.between(start, Instant.now());
+    if (response instanceof Error) {
+      Error error = (Error) response;
+      if (error.getData() instanceof Map) {
+        Map<Object, Object> builder = new HashMap<>((Map<Object, Object>) error.getData());
+        builder.put("request_duration_millis", duration.toMillis());
+        error = rpc.error(error.getCode(), error.getMessage(), builder);
+      }
+      return rpc.response(error, id);
+    } else if (response instanceof Map) {
+      Map<Object, Object> updatedResponse = new HashMap<>((Map<Object, Object>) response);
+      updatedResponse.put("request_duration_millis", duration.toMillis());
+      return rpc.response(updatedResponse, id);
+    } else {
+      return rpc.response(response, id);
+    }
+  }
+
+  private Object dispatch(Map request) {
+    if (request.get(ID) == null) {
+      return rpc.error(INVALID_REQUEST, "Missing/Invalid id");
+    }
+
+    if (!request.containsKey(METHOD)) {
+      return rpc.error(
+          METHOD_NOT_FOUND,
+          "Unknown method - use \"method\": \"list-methods\" to see available methods");
+    }
+
+    if (!(request.get(METHOD) instanceof String)) {
+      return rpc.error(
+          METHOD_NOT_FOUND,
+          "Unknown method - use \"method\": \"list-methods\" to see available methods");
+    }
+    String method = (String) request.get(METHOD);
+
+    RpcMethod target = methods.get(method);
+    if (target == null) {
+      return rpc.error(
+          METHOD_NOT_FOUND,
+          "Unknown method - use \"method\": \"list-methods\" to see available methods");
+    }
+
+    if (!(request.get(PARAMS) instanceof Map)) {
+      return rpc.error(INVALID_PARAMS, "params were not a map");
+    }
+    Map<String, Object> params = (Map) request.get(PARAMS);
+
+    try {
+      return target.apply(params);
```

Comment:
Not sure exactly what you mean here. The `Error` interface is only for typing check to differentiate error responses from result responses (since jsonrpc says you need to have either a result field or an error field). Each error object should be unique and detailed with regard to what error happened.